### PR TITLE
Support for ALPN-01 Challenge

### PR DESCRIPTION
**Pull request by ibrt**
As discussed in https://github.com/smallstep/certificates/issues/178.

This mostly follows the [existing implementation](https://github.com/letsencrypt/boulder/blob/master/va/tlsalpn.go) from Boulder with one exception: to keep it simple, instead of explicitly resolving the domain and trying to connect to all corresponding IP addresses, I let the Go standard library use its default behavior. It is still [RFC-compliant](https://tools.ietf.org/html/draft-ietf-acme-tls-alpn-01) (item 3.2), just maybe a bit less likely to succeed in some edge cases.

Besides updating the unit tests, I've also manually tested this by 1) building step-ca and running locally, 2) adding `test.local` to my hosts file, and 3) running `sudo lego --accept-tos --server="https://localhost:8443/acme/acme/directory" --email="..." --domains="test.local" --tls run`.

### CODE REVIEW

File: acme/challenge.go [hunk 1]
```diff
@@ -344,6 +358,148 @@ func (hc *http01Challenge) validate(db nosql.DB, jwk *jose.JSONWebKey, vo valida
...
+		if err = tc.storeError(db,
+			RejectedIdentifierErr(errors.Errorf("incorrect certificate for tls-alpn-01 challenge: "+
+				"leaf certificate must contain a single DNS name, %v", tc.Value))); err != nil {
+			return nil, err
+		}
+		return tc, nil
+	}
+
+	idPeAcmeIdentifier := asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 31}
+	idPeAcmeIdentifierV1Obsolete := asn1.ObjectIdentifier{1, 3, 6, 1, 5, 5, 7, 1, 30, 1}
```
Comment (cpu) [inline acme/challenge.go]: It makes me a little bit sad to see this obsolete OID propagate further. Is backwards compatibility with incorrect TLS-ALPN-01 challenge responses important for Smallstep? The larger ACME ecosystem might benefit from a nudge in the right direction by taking a firmer stance in new code and only expecting `idPeAcmeIdentifier`.

For reference, the Let's Encrypt metrics show Boulder sees the obsolete OID ~25-30% of the time. Not an insubstantial amount, but also not a majority.

File: acme/challenge.go [hunk 2]
```diff
@@ -344,6 +358,148 @@ func (hc *http01Challenge) validate(db nosql.DB, jwk *jose.JSONWebKey, vo valida
...
+			return nil, err
+		}
+		return tc, nil
+	}
+	defer conn.Close()
+
+	cs := conn.ConnectionState()
+	certs := cs.PeerCertificates
+
+	if len(certs) == 0 {
```
Comment (ibrt) [inline acme/challenge.go]: Ok this is mildly frustrating - I don't want to remove this check but I don't see a way to trigger it in tests using the HTTPS server from the standard library... Any ideas welcome

File: acme/challenge.go (hunk 1 above)
Comment (ibrt) [reply acme/challenge.go]: I don't have an opinion - if you are going to be the one merging this PR (hopefully) and prefer me to remove it I'm happy to oblige, just let me know. 

It would seem wise though to still check for this value and return a descriptive error if validation fails because of it, WDYT?

Comment (dcow) [reply acme/challenge.go]: Talked with @mmalone rather briefly, so apologies if we flip-flop, but we're cool with removing support in favor of a helpful error message, at least initially. If it's too frustrating for people, it's not hard at all to add back. Thanks @cpu for the critical eye.

Comment (ibrt) [reply acme/challenge.go]: No problem, fixed it.

Comment (ibrt) [reply acme/challenge.go]: So one interesting thing I just found out working on this is that the [golang.org/x/crypto/acme](https://github.com/golang/crypto/blob/master/acme/acme.go#L770) package is still generating certs with the obsolete identifier... I'm going to see if it can be updated :)

File: acme/challenge.go (hunk 2 above)
Comment (maraino) [reply acme/challenge.go]: @ibrt I don't think there's a way to trigger this in tests using the standard server, you might need to go very low level, and the tls client might crash before getting here.
Right now you can configure `tlsDial`, it's very possible that returning just a  `&tls.Conn{}` or perhaps a closed connection might do it. In any case, it's better to be safe always, so I wouldn't remove this check.

File: acme/challenge.go (hunk 1 above)
Comment (maraino) [reply acme/challenge.go]: @ibrt golang team welcomes contributions, you will need to create an issue first in golang/go, if it's not there yet, and then submit a PR, using GitHub, but by personal experience, I would recommend you to use directly Gerrit.
https://golang.org/doc/contribute.html

Comment (ibrt) [reply acme/challenge.go]: They actually had a pending PR for that which has now been merged: https://github.com/golang/crypto/pull/91 :)

File: acme/challenge.go (hunk 2 above)
Comment (ibrt) [reply acme/challenge.go]: The latter was a good hint - I got it to work thank you.

File: acme/challenge_test.go [hunk 3]
```diff
@@ -359,6 +446,32 @@ func TestChallengeUnmarshal(t *testing.T) {
...
+			assert.FatalError(t, err)
+			return test{
+				chb: b,
+				err: ServerInternalErr(errors.New("unexpected challenge type foo")),
+			}
+		},
+		"fail/unexpected-type-dns": func(t *testing.T) test {
+			dnsCh, err := newDNSCh()
+			assert.FatalError(t, err)
+			_tlsALPNCh, ok := dnsCh.(*dns01Challenge)
```
Comment (dopey) [inline acme/challenge_test.go]: Should the names here be `_dnsCh`, etc? (assuming alpn is here by mistake)

File: acme/challenge_test.go [hunk 4]
```diff
@@ -359,6 +446,32 @@ func TestChallengeUnmarshal(t *testing.T) {
...
+			assert.Fatal(t, ok)
+			_tlsALPNCh.baseChallenge.Type = "foo"
+			b, err := json.Marshal(dnsCh)
+			assert.FatalError(t, err)
+			return test{
+				chb: b,
+				err: ServerInternalErr(errors.New("unexpected challenge type foo")),
+			}
+		},
```
Comment (dopey) [inline acme/challenge_test.go]: Can you add a `ok/tls-alpn01` challenge? Should be as easy as copy pasta-ing `ok/http` and just changing some of the var names.

File: acme/challenge_test.go (hunk 3 above)
Comment (ibrt) [reply acme/challenge_test.go]: Yup my bad

File: acme/challenge_test.go (hunk 4 above)
Comment (ibrt) [reply acme/challenge_test.go]: Done

**Comment by ibrt** [comment]
It should be good now, thanks for reviewing!
